hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java

```diff
@@ -235,123 +231,22 @@ public FileEncryptionInfo getFileEncryptionInfo() {
     return encInfo;
   }
 
-  public List<OzoneAclInfo> getAcls() {
+  public List<OzoneAcl> getAcls() {
```

Comment:
We can leave this as OzoneAclInfo, and make OmBucketInfo OzoneAcl -> OzoneAclInfo.
In this way, we can avoid conversion of OzoneAcl list during OM operations.
And this will be useful in HA code path, as we operate directly OMRequest objects. (I mean here we get the data from OMRequest, and perform operation and store in cache)

Reply:
OzoneAcl class is introduced to make acl operation such as add/remove/set acl consistent. Unfortunately, some of the early acl implementation uses OzoneAclInfo (protobuf class) directly, which this ticket attempts to unify. 

The protobuf class OzoneAclInfo is good for maintain compatibility for RPC and DB persistence. As a result, we will only do the protobuf conversion in those places.

Reply:
What I mean here is:
Take an example OMBucketAddAclRequest
```
  public OMBucketAddAclRequest(OMRequest omRequest) {
    super(omRequest, bucketAddAclOp);
    OzoneManagerProtocolProtos.AddAclRequest addAclRequest =
        getOmRequest().getAddAclRequest();
    path = addAclRequest.getObj().getPath();
    ozoneAcls = Lists.newArrayList(
        OzoneAcl.fromProtobuf(addAclRequest.getAcl()));
```

If OMBucketInfo, has direct protobuf structures we can avoid the conversion.

Now, we do 1 conversion in constructor, and the other when write to DB. ozoneAcl -> proto -> byte Array. If we can have direct protobuf structures, we can save unnecessary protobuf conversions.

Reply:
> As a result, we will only do the protobuf conversion in those places.

Not sure if I am missing something here. From the above example, we are doing the conversion in 2 places.

Reply:
1st conversion is for RPC based request.
2nd conversion is for persistence to DB. 
 
In between, we use OzoneAcl to manipulate the ACL. I understand we could save the conversion by using the protobuf class directly. Here are reasons I choose to keep Java class instead:

1. Raw protobuf class can't be extended easily for acl operations. We will have to to wrap the acl operations anyway. 

2. Given the scatter logic of ACL ops for volume/bucket/key/prefix, I would prefer to have a unified code base that can be maintained/tested easily in the long run.

3. ACL write operations are far less than ACL read. I would not expect this to have big impact to ozone perf.

Reply:
`Raw protobuf class can't be extended easily for acl operations. We will have to to wrap the acl operations anyway.`
Not understood what do you mean by wrap ACL operations anyway.

`Given the scatter logic of ACL ops for volume/bucket/key/prefix, I would prefer to have a unified code base that can be maintained/tested easily in the long run.`

I understand that instead of spreading logic across in multiple places keep acl logic in single place, which can be used in all bucket/key/prefix ops. I agree with that point. But I have not understood why same cannot be done with protobuf acl objects.

3. Agreed thar write acl operations/ will be lesser than read Acls.

But if you think this is the right way to go, I am fine with it.

Reply:
And also this will not only help acls But bucket/key creation also, as now Bucket/KeyInfo will have proto buf -> internal Ozone object. This also can be avoided. (So, when each key creation I don't need to convert acls set during the creation of key from proto to OzoneAcl Objects.

Reply:
bq. Raw protobuf class can't be extended easily for acl operations. We will have to to wrap the acl operations anyway.

This is based on [protobuf doc](https://developers.google.com/protocol-buffers/docs/javatutorial) below. My understanding is that protobuf is only good for serialization/transport. Our acl operation specific logic should be wrapped without dependency on it so that it can be maintained easily for the long term.  
 
"Protocol buffer classes are basically dumb data holders (like structs in C); they don't make good first class citizens in an object model. If you want to add richer behavior to a generated class, the best way to do this is to wrap the generated protocol buffer class in an application-specific class."

bq. And also this will not only help acls But bucket/key creation also, as now Bucket/KeyInfo will have proto buf -> internal Ozone object. This also can be avoided. (So, when each key creation we don't need to convert acls set during the creation of key from proto to OzoneAcl Objects.)

This requires a broader change in the interface itself, which is beyond the scope of this JIRA, i.e. consolidate the core acl ops.

Reply:
Thank You for info. Here we are only modifying acl depending on add/remove/set Acl. I think for this we can perform on protobuf structures. Like how it was done before in OmKeyInfo.java (add/remove/setAcl). As keeping in this way, we will avoid unnecessary protobuf conversions.
If you feel this is the way to do, I am okay with it.

Reply:
Good point. CheckAccess does the the byte array based protobuf acl to BitSet based acl conversion each time, which is most of the work we do for protobuf conversion from OzoneAclInfo to OzoneAcl. Consolidate this into OzoneAcl class allows us to change the logic in a single place for optimization in future. 

Key creation usually needs to inherit the acls from prefix tree, where efficient merge of supplied acls along with the acls from in-memory prefix tree. It is currently not done properly tracked by HDDS-1913.

Reply:
> Good point. CheckAccess does the the byte array based protobuf acl to BitSet based acl conversion each time, which is most of the work we do for protobuf conversion from OzoneAclInfo to OzoneAcl. Consolidate this into OzoneAcl class allows us to change the logic in a single place for optimization in future.
Ya right. I think it should be fine I believe. (But still, we have little advantage of not converting entire protobuf object. But I am fine with the current way. :))

> Key creation usually needs to inherit the acls from prefix tree, where efficient merge of supplied acls along with the acls from in-memory prefix tree. It is currently not done properly tracked by HDDS-1913.

HDDS-1913 is for fixing OzoneBucket and RpcClient APIS for acl. Is there another jira to fix the above issue?

Reply:
Similar issue can be found in the PR of HDDS-1909 and HDDS-1975. 
The new helper OzoneAclUtil.inheritDefaultAcls should guarantee they are merged properly.

Reply:
HDDS-1975 is opened for inherit default ACL for HA code path. I can rebase the PR once this went in to use this new method.

Reply:
Sounds good to me. Thanks!